File: simpledb/src/main/scala/simpledb.scala
package aws.simpledb

import java.util.Date

import scala.annotation.implicitNotFound
import scala.util.{ Try, Success }
import scala.concurrent.Future
import scala.concurrent.ExecutionContext.Implicits.global
import play.api.libs.ws._

import aws.core._
import aws.core.Types._
import aws.core.parsers._
import aws.core.signature.V2
import aws.simpledb.SDBParsers._

case class SDBAttribute(name: String, value: String, replace: Option[Boolean] = None)
case class SDBItem(name: String, attributes: Seq[SDBAttribute])
case class SDBDomain(name: String)
case class SDBDomainMetadata(
  timestamp: Date,
  itemCount: Long,
  attributeValueCount: Long,
  attributeNameCount: Long,
  itemNamesSizeBytes: Long,
  attributeValuesSizeBytes: Long,
  attributeNamesSizeBytes: Long)

case class SimpleDBMeta(requestId: String, boxUsage: String) extends Metadata

object SDBRegion {

  val NAME = "sdb"

  val US_EAST_1 = AWSRegion.US_EAST_1(NAME)
  val US_WEST_1 = AWSRegion.US_WEST_1(NAME)
  val US_WEST_2 = AWSRegion.US_WEST_2(NAME)
  val EU_WEST_1 = AWSRegion.EU_WEST_1(NAME)
  val ASIA_SOUTHEAST_1 = AWSRegion.ASIA_SOUTHEAST_1(NAME)
  val ASIA_NORTHEAST_1 = AWSRegion.ASIA_NORTHEAST_1(NAME)
  val SA_EAST_1 = AWSRegion.SA_EAST_1(NAME)

  implicit val DEFAULT = US_EAST_1
}

object SimpleDB {

  object Parameters {
    def DomainName(a: String) = ("DomainName" -> a)
    def ItemName(a: String) = ("ItemName" -> a)
    def MaxNumberOfDomains(n: Int) = ("MaxNumberOfDomains" -> n.toString)
    def NextToken(n: String) = ("NextToken" -> n)
    def ConsistentRead(c: Boolean) = ("ConsistentRead" -> (if (c) "true" else "false"))
    def AttributeName(n: String) = ("AttributeName" -> n)
    def Attributes(attrs: Seq[SDBAttribute]): Seq[(String, String)] = (for ((attribute, i) <- attrs.zipWithIndex) yield {
      Seq(
        "Attribute.%d.Name".format(i + 1) -> attribute.name,
        "Attribute.%d.Value".format(i + 1) -> attribute.value) ++ {
          if (attribute.replace.isDefined) Seq("Attribute.%s.Replace".format((i + 1).toString) -> attribute.replace.get.toString) else Nil
        }
    }).flatten
    def Items(items: Seq[SDBItem]): Seq[(String, String)] = (for ((item, i) <- items.zipWithIndex) yield {
      val prefix = "Item.%d.".format(i + 1)
      (prefix + "ItemName" -> item.name) +: Attributes(item.attributes).map { attr =>
        (prefix + attr._1 -> attr._2)
      }
    }).flatten
    def SelectExpression(expression: String) = ("SelectExpression" -> expression)
  }

  import AWS.Parameters._
  import Parameters._

  private def request(parameters: (String, String)*)(implicit region: AWSRegion): Future[Response] = {
    WS.url("https://" + region.host + "/?" + V2.signedUrl("GET", parameters)).get()
  }

  private def tryParse[M <: Metadata, T](resp: Response)(implicit p: Parser[Result[M, T]]) = Parser.parse[Result[M, T]](resp).fold(
    e => throw new RuntimeException(e),
    identity)

  /**
   * This helper will fetch data using request, and try to parse the response to the specified types
   */
  private def get[M <: Metadata, T](parameters: (String, String)*)(implicit region: AWSRegion, p: Parser[Result[M, T]]): Future[Result[M, T]] =
    request(parameters: _*).map(tryParse[M, T])

  /**
   * Creates a domain with the given name
   */
  def createDomain(domainName: String)(implicit region: AWSRegion): Future[EmptyResult[SimpleDBMeta]] =
    get[SimpleDBMeta, Unit](Action("CreateDomain"), DomainName(domainName))

  /**
   * Deletes the given domain
   */
  def deleteDomain(domainName: String)(implicit region: AWSRegion): Future[EmptyResult[SimpleDBMeta]] =
    get[SimpleDBMeta, Unit](Action("DeleteDomain"), DomainName(domainName))

  /**
   * Lists domains starting with the nextToken, if present, and giving the max number of domains given
   */
  def listDomains(maxNumberOfDomains: Int = 100, nextToken: Option[String] = None)(implicit region: AWSRegion): Future[Result[SimpleDBMeta, Seq[SDBDomain]]] = {
    val params = Seq(
      Action("ListDomains"),
      MaxNumberOfDomains(maxNumberOfDomains)) ++ nextToken.map(NextToken(_)).toSeq

    get[SimpleDBMeta, Seq[SDBDomain]](params: _*)
  }

  /**
   * Puts the attributes into SimpleDB
   */
  def putAttributes(domainName: String,
                    itemName: String,
                    attributes: Seq[SDBAttribute])(implicit region: AWSRegion): Future[EmptyResult[SimpleDBMeta]] = {
    val params = Seq(
      Action("PutAttributes"),
      DomainName(domainName),
      ItemName(itemName)) ++ Attributes(attributes)

    get[SimpleDBMeta, Unit](params: _*)
  }

  /**
   * Delete all or some attributes for an item. To delete the whole item, pass its name with an empty attribute list.
   */
  def deleteAttributes(domainName: String, item: SDBItem)(implicit region: AWSRegion): Future[EmptyResult[SimpleDBMeta]] = {
    val params = Seq(
      Action("DeleteAttributes"),
      DomainName(domainName),
      ItemName(item.name)) ++ Attributes(item.attributes)

    get[SimpleDBMeta, Unit](params: _*)
  }

  /**
   * Gets one or all Attributes from the given domain name and item name
   */
  def getAttributes(domainName: String,
                    itemName: String,
                    attributeName: Option[String] = None,
                    consistentRead: Boolean = false)(implicit region: AWSRegion): Future[Result[SimpleDBMeta, Seq[SDBAttribute]]] = {
    val params = Seq(
      Action("GetAttributes"),
      DomainName(domainName),
      ItemName(itemName),
      ConsistentRead(consistentRead)) ++ attributeName.map(AttributeName(_)).toSeq

    get[SimpleDBMeta, Seq[SDBAttribute]](params: _*)
  }

  /**
   * Get detailed information about a domain
   */
  def domainMetadata(domainName: String)(implicit region: AWSRegion): Future[Result[SimpleDBMeta, SDBDomainMetadata]] =
    get[SimpleDBMeta, SDBDomainMetadata](Action("DomainMetadata"), DomainName(domainName))

  /**
   * Returns a set of Attributes for ItemNames that match the select expression (similar to SQL)
   */
  def select(expression: String, nextToken: Option[String] = None, consistentRead: Boolean = false)(implicit region: AWSRegion): Future[Result[SimpleDBMeta, Seq[SDBItem]]] = {
    val params = Seq(
      Action("Select"),
      SelectExpression(expression),
      ConsistentRead(consistentRead)) ++ nextToken.map(NextToken(_)).toSeq

    get[SimpleDBMeta, Seq[SDBItem]](params: _*)
  }

  /**
   * Put attributes for more than one item
   */
  def batchPutAttributes(domainName: String, items: Seq[SDBItem])(implicit region: AWSRegion): Future[EmptyResult[SimpleDBMeta]] = {
    val params = Seq(
      Action("BatchPutAttributes"),
      DomainName(domainName)) ++ Items(items)
    get[SimpleDBMeta, Unit](params: _*)
  }

  /**
   * Delete attributes for more than one item
   */
  def batchDeleteAttributes(domainName: String, items: Seq[SDBItem])(implicit region: AWSRegion): Future[EmptyResult[SimpleDBMeta]] = {
    val params = Seq(
      Action("BatchDeleteAttributes"),
      DomainName(domainName)) ++ Items(items)
    get[SimpleDBMeta, Unit](params: _*)
  }

}

File: simpledb/src/main/java/com/pellucid/aws/simpledb/SimpleDB.java
package com.pellucid.aws.simpledb;

import java.util.List;
import java.util.ArrayList;

import play.libs.F;
import play.libs.Scala;
import scala.collection.JavaConversions;
import scala.collection.Seq;
import scala.concurrent.Future;
import scala.runtime.BoxedUnit;

import com.pellucid.aws.internal.AWSJavaConversions;
import com.pellucid.aws.internal.MetadataConvert;
import com.pellucid.aws.results.Result;

public class SimpleDB {

    private final aws.simpledb.SDBRegion scalaRegion;

    public SimpleDB(SDBRegion region) {
        this.scalaRegion = SimpleDB.scalaRegion(region);
    }

    /**
     * Creates a new domain. The domain name must be unique among the domains associated with the Access Key ID used.
     * The CreateDomain operation might take 10 or more seconds to complete.
     * CreateDomain is an idempotent operation; running it multiple times using the same domain name will not result
     * in an error response.
     *
     * You can create up to 250 domains per account.
     * If you require additional domains, go to http://aws.amazon.com/contact-us/simpledb-limit-request/.
     *
     * @param domainName
     */
    public F.Promise<Result<SimpleDBMeta, Object>> createDomain(String domainName) {
        return convertEmptyResult(aws.simpledb.SimpleDB.createDomain(domainName, scalaRegion));
    }

    /**
     * Deletes the given domain. Any items (and their attributes) in the domain are deleted as well.
     * The DeleteDomain operation might take 10 or more seconds to complete.
     *
     * Running DeleteDomain on a domain that does not exist or running the function multiple times
     * using the same domain name will not result in an error response.
     *
     * @param domainName
     */
    public F.Promise<Result<SimpleDBMeta, Object>> deleteDomain(String domainName) {
        return convertEmptyResult(aws.simpledb.SimpleDB.deleteDomain(domainName, scalaRegion));
    }

    public F.Promise<Result<SimpleDBMeta, List<String>>> listDomains() {
        return listDomains(100, null);
    }

    public F.Promise<Result<SimpleDBMeta, List<String>>> listDomains(Integer maxNumberOfDomains) {
        return listDomains(maxNumberOfDomains, null);
    }

    public F.Promise<Result<SimpleDBMeta, List<String>>> listDomains(String nextToken) {
        return listDomains(100, nextToken);
    }

    /**
     * Lists all domains associated with the Access Key ID.
     * It returns domain names up to the limit set by MaxNumberOfDomains. A NextToken is returned if
     * there are more than MaxNumberOfDomains domains. Calling ListDomains successive times with the NextToken returns up to MaxNumberOfDomains more domain names each time.
     *
     * @param maxNumberOfDomains limit the response to a size
     * @param nextToken: optionally provide this value you got from a previous request to get the next page
     */
    public F.Promise<Result<SimpleDBMeta, List<String>>> listDomains(Integer maxNumberOfDomains, String nextToken) {
        return AWSJavaConversions.toResultPromise(aws.simpledb.SimpleDB.listDomains(maxNumberOfDomains, Scala.Option(nextToken), scalaRegion),
                new MetadataConvert(),
                new F.Function<Seq<aws.simpledb.SDBDomain>, List<String>>() {
                    @Override public List<String> apply(Seq<aws.simpledb.SDBDomain> domains) {
                        return domainSeqToStringList(domains);
                    }
        });
    }

    private static List<String> domainSeqToStringList(Seq<aws.simpledb.SDBDomain> domains) {
        List<String> result = new ArrayList<String>();
        for (aws.simpledb.SDBDomain domain: JavaConversions.seqAsJavaList(domains)) {
            result.add(domain.name());
        }
        return result;
    }

    private static F.Promise<Result<SimpleDBMeta, Object>> convertEmptyResult(Future<aws.core.Result<aws.simpledb.SimpleDBMeta, BoxedUnit>> scalaResult) {
        return AWSJavaConversions.toResultPromise(scalaResult, new MetadataConvert(), new F.Function<BoxedUnit, Object>() {
            @Override public Object apply(BoxedUnit unit) throws Throwable {
                return null;
            }
        });
    }

    private static aws.simpledb.SDBRegion scalaRegion(SDBRegion region) {
        switch (region) {
        case US_EAST_1: return aws.simpledb.SDBRegion$.MODULE$.US_EAST_1();
        case US_WEST_1: return aws.simpledb.SDBRegion$.MODULE$.US_WEST_1();
        case US_WEST_2: return aws.simpledb.SDBRegion$.MODULE$.US_WEST_2();
        case EU_WEST_1: return aws.simpledb.SDBRegion$.MODULE$.EU_WEST_1();
        case ASIA_SOUTHEAST_1: return aws.simpledb.SDBRegion$.MODULE$.ASIA_SOUTHEAST_1();
        case ASIA_NORTHEAST_1: return aws.simpledb.SDBRegion$.MODULE$.ASIA_NORTHEAST_1();
        case SA_EAST_1: return aws.simpledb.SDBRegion$.MODULE$.SA_EAST_1();
        }
        return aws.simpledb.SDBRegion$.MODULE$.DEFAULT();
    }

}

